ext/git.c

```diff
@@ -115,104 +137,88 @@ void removeCredentials(zend_string* repo_url) {
-            Z_TYPE_P(git_repository_url) == IS_STRING) {
-            success = add_git_info(meta, Z_STR_P(git_commit_sha), Z_STR_P(git_repository_url), is_root_span, true);
-        }
-    }
-
-    if (success && is_root_span) {
-        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.commit.sha"));
-        zend_hash_str_del(Z_ARR_P(meta), ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && ZSTR_LEN(commit_sha) > 0 && ZSTR_LEN(repository_url) > 0) {
+        remove_credentials(repository_url);
+        add_git_info(carrier, commit_sha, repository_url);
+        success = true;
+        zend_string_release(git_dir);
     }
 
     return success;
 }
 
-bool inject_from_git_files(zval* meta, bool is_root_span) {
-    char cwd[PATH_MAX];
-    if (!getcwd(cwd, sizeof(cwd))) {
-        return false;
-    }
-
-    char git_dir[PATH_MAX];
-    snprintf(git_dir, sizeof(git_dir), "%s/.git", cwd);
-
-    zend_string* git_commit_sha = get_commit_sha(git_dir);
-    zend_string* git_repository_url = get_repository_url(git_dir);
+void ddtrace_inject_git_metadata(zval *git_metadata_zv) {
+    zend_string *cwd_zstr = NULL;
 
-    if (!git_commit_sha || !git_repository_url) {
-        if (git_commit_sha) zend_string_release(git_commit_sha);
-        if (git_repository_url) zend_string_release(git_repository_url);
-        return false;
+    if (SG(options) & SAPI_OPTION_NO_CHDIR) {
+        const char *script_filename = SG(request_info).path_translated;
+        cwd_zstr = zend_string_init(script_filename, strlen(script_filename), 1);
+    } else {
+        char cwd[PATH_MAX];
+        if (getcwd(cwd, sizeof(cwd)) == NULL) {
+            return;
+        }
+        cwd_zstr = zend_string_init(cwd, strlen(cwd), 1);
     }
 
-    bool result = add_git_info(meta, git_commit_sha, git_repository_url, is_root_span, true);
-
-    zend_string_release(git_commit_sha);
-    zend_string_release(git_repository_url);
-
-    return result;
-}
+    zval *entry = zend_hash_find(&DDTRACE_G(git_metadata), cwd_zstr);
 
-void ddtrace_inject_git_metadata(zval* meta, bool is_root_span) {
-    ddtrace_git_metadata git_metadata = DDTRACE_G(git_metadata);
-    if (git_metadata.called_once) {
-        add_git_info(meta, git_metadata.commit_sha, git_metadata.repository_url, is_root_span, false);
+    if (entry && Z_TYPE_P(entry) == IS_FALSE) {
+        zend_string_release(cwd_zstr);
         return;
     }
 
-
-    if (inject_from_env(meta, is_root_span) ||
-        inject_from_global_tags(meta, is_root_span) ||
-        inject_from_git_files(meta, is_root_span)) {
+    if (entry) {
+        zend_string_release(cwd_zstr);
+        ZVAL_COPY(git_metadata_zv, entry);
         return;
     }
+
+    if (inject_from_env(git_metadata_zv) || inject_from_global_tags(git_metadata_zv) || inject_from_git_dir(git_metadata_zv, cwd_zstr)) {
+        zend_hash_add(&DDTRACE_G(git_metadata), cwd_zstr, git_metadata_zv);
```

Comment:
It's highly forbidden to persist objects across requests. You may only persist raw strings across a request.